KAFKA-9047: Making AdminClient group operations respect retries and backoff

skaundinya15 opened the pull request:
https://issues.apache.org/jira/browse/KAFKA-9047

Previously, `AdminClient` group operations did not respect a `Call`'s number of configured tries and retry backoff. This could lead to tight retry loops that put a lot of pressure on the broker. This PR introduces fixes that ensures for all group operations the `AdminClient` respects the number of tries and the backoff a given `Call` has.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
added new constructor with optional arguments

Commit:
Merge branch 'trunk' into KAFKA-9047

Commit:
Merge branch 'trunk' into KAFKA-9047

Commit:
wrote initial test to test retrybackoff

Commit:
made changes to get retry ms logic

Commit:
made changes to deadline logic

Commit:
Merge branch 'trunk' into KAFKA-9047

Commit:
added another condition to retry backoff for offset commit test

Commit:
initial version of retry test and logic

Commit:
better placement of num retry logic

Commit:
adding updated logic to other calls

Commit:
added tests for DescribeConsumerGroup

Commit:
small tweak to test

Commit:
added tests for ListConsumerGroupOffsets

Commit:
fixing testOffsetCommitRetryBackoff test

Commit:
fixing testDescribeConsumerGroupRetryBackoff

Commit:
fixing ListConsumerGroupOffsetsRetryBackoff

Commit:
fixing testDeleteConsumerGroupsRetryBackoff

Commit:
adding tests for DeleteConsumerGroupOffsets

Commit:
adding tests for RemoveMembersFromGroupRetry

Commit:
removed changes for listOffsets

Commit:
added new max retries reached exception

Commit:
Merge branch 'trunk' into KAFKA-9047

Commit:
Merge branch 'trunk' into KAFKA-9047

Commit:
Merge branch 'trunk' into KAFKA-9047

Commit:
Merge branch 'trunk' into KAFKA-9047

Commit:
changed naming error

Commit:
removed stray comma

Commit:
fixed misnaming

Commit:
fixed misnaming

hachikuji commented:
ok to test

hachikuji commented on clients/src/main/java/org/apache/kafka/common/errors/MaxRetriesReachedException.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Indicates that a request has reached the maximum number of tries.
+ */
+public class MaxRetriesReachedException extends KafkaException {
```
I think we can just use TimeoutException. We would need a KIP to expose a new exception type.

skaundinya15 replied:
Ah okay wasn't aware of that. Will make the change to `TimeoutException`.

Commit:
removed exception and fixed checkstlye issues

skaundinya15 commented:
@hachikuji Thanks for the initial review! I've addressed your comments and pushed the changes. It's ready for another round of reviews whenever you are.

ijuma commented:
ok to test

Commit:
fixing checkstyle issues

ctan888 commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2813,7 +2838,10 @@ void handleResponse(AbstractResponse abstractResponse) {
                 if (ConsumerGroupOperationContext.hasCoordinatorMoved(response)) {
-                    rescheduleFindCoordinatorTask(context, () -> getDescribeConsumerGroupsCall(context));
+                    long nextAllowedTryMs = calculateNextAllowedRetryMs(this.nextAllowedTryMs());
+                    Call call = getDescribeConsumerGroupsCall(context, Optional.of(this.tries() + 1),
```
I'm just curious if we can call Call::fail to push the call instance into the queue again?

skaundinya15 replied:
We won't be able to do this because we need to be able to reschedule the `findCoordinator` call and then subsequently schedule the call. It's easier to pass the state across a new construction of the call and have the fail logic when we are in the `findCoordinator` logic.

soondenana commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -690,6 +690,13 @@ public Node provide() {
+            Optional<Integer> numTries, Optional<Long> nextAllowedTryMs) {
+            this(false, callName, deadlineMs, nodeProvider);
+            this.tries = numTries.orElse(0);
+            this.nextAllowedTryMs = nextAllowedTryMs.orElse(0L);
```
nit: add some sanity check on these numbers (like should be non-negative etc).

Also update `toString` method to include this information.

skaundinya15 replied:
Done.

soondenana commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2683,6 +2698,14 @@ private void rescheduleFindCoordinatorTask(ConsumerGroupOperationContext<?, ?> c
+        Call call = nextCall.get();
+
+        if (call.tries() > maxRetries) {
+            log.debug("Max retries for {} reached", call);
```
nit: Maybe update the message to "Max retries maxRetries for <call> reached."

skaundinya15 replied:
Done.

soondenana commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -690,6 +690,13 @@ public Node provide() {
         }
 
+        Call(String callName, long deadlineMs, NodeProvider nodeProvider,
+            Optional<Integer> numTries, Optional<Long> nextAllowedTryMs) {
```
Not sure if we need to make these `Optional`. `0` seems to be a good default value for these.

skaundinya15 replied:
I initially thought that we might need one, because we wouldn't know, but after thinking about your comment that `0` is a good default value, I have changed it to reflect as such.

soondenana commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -3535,6 +3576,12 @@ void handleFailure(Throwable throwable) {
+    private long calculateNextAllowedRetryMs(long currNextAllowedRetryMs) {
+        return (currNextAllowedRetryMs == 0) ?
+            time.milliseconds() + retryBackoffMs :
+            currNextAllowedRetryMs + retryBackoffMs;
```
To keep this logic same as one in `Call::fail` method, lets set the new time as:

> nextAllowedTryMs = now + retryBackoffMs

skaundinya15 replied:
Done.

soondenana reviewed (commented):
Thanks for fixing this. The logic looks good overall. I have left few comments.

Commit:
Merge branch 'trunk' into KAFKA-9047

Commit:
addressing PR comments

skaundinya15 commented:
Thanks for the reviews @soondenana! I have addressed your feedback and pushed changes just now. Ready for another round of reviews when you are.

hachikuji commented:
retest this please

hachikuji commented:
ok to test

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -690,6 +690,15 @@ public Node provide() {
         }
 
+        Call(String callName, long deadlineMs, NodeProvider nodeProvider,
+            int numTries, long nextAllowedTryMs) {
```
nit: this alignment is a little weird. Typically we would align with the arg above

skaundinya15 replied:
Ack, will fix.

soondenana commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -691,10 +691,12 @@ public Node provide() {
-            this.tries = numTries.orElse(0);
-            this.nextAllowedTryMs = nextAllowedTryMs.orElse(0L);
+            assert numTries >= 0;
+            assert nextAllowedTryMs >= 0;
```
I am not sure we enable java asserts when running Kafka server. Lets check the condition and throw `IllegalArgumentException` instead.

skaundinya15 replied:
Done.

soondenana commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -3576,10 +3574,8 @@ void handleFailure(Throwable throwable) {
-        return (currNextAllowedRetryMs == 0) ?
-            time.milliseconds() + retryBackoffMs :
-            currNextAllowedRetryMs + retryBackoffMs;
+    private long calculateNextAllowedRetryMs(long now) {
```
nit: As we always pass same value to this method, we can drop the parameter and use `time.milliseconds` within this method.

skaundinya15 replied:
Done.

soondenana reviewed (approved):
LGTM.

Couple of nits more. Nice tests.

Commit:
addressing PR comments

Commit:
fixing type of exception thrown

skaundinya15 commented:
Thanks for +1 @soondenana! @hachikuji whenever you're ready, the PR is ready for a review from you.

ctan888 commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2683,6 +2702,14 @@ private void rescheduleFindCoordinatorTask(ConsumerGroupOperationContext<?, ?> c
+
+        Call call = nextCall.get();
+
+        if (call.tries() > maxRetries) {
```
Is it possible to move the check to the AdminClientRunnable::call / enqueue to make this code pieces reusable?

ctan888 replied:
What about adding a new class method in Call?

skaundinya15 replied:
We could but we still have to do a check as to whether or not we failed the call to make sure we return from the method. It would still produce the same amount of code and I'm not sure if it would be any more efficient.

skaundinya15 replied:
And I'm not sure if we should move the check to `AdminClientRunnable::call/enqueue` because this is pretty specific to the consumer groups API. However if we see there's a potential reuse of code elsewhere we can definitely consider doing so.

Commit:
Merge branch 'trunk' into KAFKA-9047

hachikuji commented:
ok to test

hachikuji commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -690,6 +690,20 @@ public Node provide() {
+            int numTries, long nextAllowedTryMs) {
+            this(false, callName, deadlineMs, nodeProvider);
+            if (numTries < 0) {
+                throw new IllegalArgumentException("Number of tries for a call cannot be negative!");
```
nit: I feel like a jerk for saying this, but can we avoid the exclamation points? The enthusiasm can be a bit vexing for an operator who hits this error.

skaundinya15 replied:
Fair point, will remove the exclamation mark.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -799,12 +813,17 @@ boolean handleUnsupportedVersionException(UnsupportedVersionException exception)
             return internal;
         }
+
+        public int tries() {
```
nit: I think technically we don't need this. For internal classes, I tend to favor accessing fields directly when the class is nested. I don't feel strongly, so up to you.

skaundinya15 replied:
I do have a preference towards setters and getters, but in this case since it is a nested class as you said, I'll remove it.

skaundinya15 replied:
Actually, since `tries` is a private field inside `Call`, I have the `tries()` getter method in order to pass the state of the number of tries across calls. In a future PR we could remove this when we have a better `CallContext` object as part of the exponential backoff KIP.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -690,6 +690,20 @@ public Node provide() {
             this(false, callName, deadlineMs, nodeProvider);
         }
 
+        Call(String callName, long deadlineMs, NodeProvider nodeProvider,
```
One of the annoying aspects of the code below is that we have a lot of redundant logic for constructing a call in the context of a previously failed call. I wonder if it would simplify the logic if we added a constructor which accepts the failed call as an argument and then adjusts `tries` etc accordingly.

skaundinya15 replied:
Yeah that was something that was bothering me as well. However when I tried to create a constructor that would take the failed call, it wasn't working as to use the constructor I also had to implement the methods in the `Call` class. That's why I think we get the call normally from the method that is returning the call. Taking the `getDescribeConsumerGroupsCall` example, it could get a bit more messy if wanted to implement the methods within the new constructor:

```java
                if (ConsumerGroupOperationContext.hasCoordinatorMoved(response)) {
                    Call call = this.call();
                    Call call2 = new Call(call) {
                        @Override
                        Builder createRequest(final int timeoutMs) {
                            return call.createRequest(timeoutMs);
                        }

                        @Override
                        void handleResponse(final AbstractResponse abstractResponse) {
                            call.handleResponse(abstractResponse);
                        }

                        @Override
                        void handleFailure(final Throwable throwable) {
                            call.handleFailure(throwable);
                        }
                    };
                    rescheduleFindCoordinatorTask(context, () -> call2);
                    return;
                }
```

With a new constructor in `Call` looking like this:

```java
        Call(Call call) {
            this.internal = call.internal;
            this.callName = call.callName;
            this.deadlineMs = call.deadlineMs;
            this.nodeProvider = call.nodeProvider;
            this.aborted = call.aborted;
            this.curNode = call.curNode;
            this.tries += 1;
            this.nextAllowedTryMs = calculateNextAllowedRetryMs();
        }
```


Thinking about this a bit more though, another way I was thinking we could simplify the logic without having to reimplement the methods for the constructor was like this (taking the same example of the `getDescribeConsumerGroupsCall`:

```java
// in the handleResponse section
if (ConsumerGroupOperationContext.hasCoordinatorMoved(response)) {
     Call call = getDescribeConsumerGroupsCall(context);
     call.updateTriesAndNextAllowedTryMs();
     rescheduleFindCoordinatorTask(context, () -> call);
     return;
}
```

and the `Call` class would get this new method:

```java
void updateTriesAndNextAllowedTryMs() {
            this.tries += 1;
            this.nextAllowedTryMs = calculateNextAllowedRetryMs();
}
```

Personally, I prefer to keep the code more readable by getting the call from the method and simply using the internal method to update it - however if you prefer we use the constructor way, I can change it to that.

hachikuji replied:
Hmm.. I see. Yeah, it seems a bit awkward. Let me offer one more suggestion before I give up. I think the awkwardness is a result of the fact that we cannot easily chain the retry context from one call to another. Suppose that we create a separate object for this:
```java
class CallContext {
  long deadlineMs;
  int tries;
  long nextAllowedTryMs;

  CallContext retry(long now) {
    return new CallContext(
      deadlineMs,
      tries + 1,
      now + backoffMs);
  }
}
```
Then inside `Call`, we would have a reference to the context rather than the individual properties. Would that work?

skaundinya15 replied:
That could work, but is there a reason we don't want to directly modify the properties in the call upon retry by having a method inside `Call` as suggested above? Something I was thinking about with   having a `CallContext` across calls is that would it be restricted to these individual properties? Or do we expect that the properties that might have to be chained across calls could increase in the future? Just trying to better understand how much scope the `CallContext` would have, considering there's already a `ConsumerGroupOperationContext` object passed across calls.

hachikuji replied:
Mainly just trying to consolidate the logic for updating the retry state so that we don't have to remember to do it for new APIs. Also, eventually we want to have exponential backoff logic. I was thinking `CallContext` would be a good way to encapsulate any additional state that we would need.

skaundinya15 replied:
Yeah that's a fair point, the only thing is that those properties are directly accessed in a lot of places (when it comes to failure and retry logic) and I'm thinking that changing the way those properties are encapsulated would require refactoring that could possibly be beyond the scope of this PR. 

For example, in the `fail` method of the `Call` class, these attributes are directly accessed many times to determine whether or not to actually fail the call or retry it. Also, I'm considering how changing the encapsulation would affect all other calls outside of the consumer groups operations, as those would also have to be updated to use the `CallContext` object instead.

The other option I'm thinking (to reduce the amount of refactoring) would be to have both `CallContext` and the properties at play, but that would be a lot of redundant information and I'm not quite sure if we'd want that. 

I can change it to be either way, but I personally think having a context for the `Call` is something that should be done so we can more easily chain it across calls, but perhaps is out of scope for this PR and could be picked up in a future PR.

hachikuji replied:
Ok, let's leave this as a potential future improvement (perhaps as part of the the exponential backoff kip).

hachikuji reviewed (commented):
Thanks for the patch. The logic looks good. I just had some small comments. I will review test cases shortly (really happy to see the coverage!).

skaundinya15 committed:
addressing PR comments

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2683,6 +2702,14 @@ private void rescheduleFindCoordinatorTask(ConsumerGroupOperationContext<?, ?> c
+
+        Call call = nextCall.get();
+
+        if (call.tries > maxRetries) {
```
I missed this on the first pass, but would there be any harm moving this call to `enqueue`? That way we do not need to repeat this check for other scenarios involving chaining or partial failures. (For example, I assume we'd end up doing the same thing in `rescheduleMetadataTask`).

We should probably also try to reuse `failWithTimeout` for a consistent exception message.

skaundinya15 replied:
I can't think of any issues with moving this logic to `enqueue`, the only reason I put this inside the `rescheduleFindCoordinatorTask` was so that we could avoid making a call to find the coordinator if the call had maxed out on retries. I will try it with this logic to see if there are any issues with moving it to `enqueue`.

Also the `fail` method already calls `failWithTimeout` when a call has maxed out the retries: https://github.com/apache/kafka/blob/5a17d56fca676c70579222b1fbe5d1dd501c9297/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L757-L759

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -3584,7 +3633,10 @@ void handleResponse(AbstractResponse abstractResponse) {
+                    long nextAllowedTryMs = calculateNextAllowedRetryMs();
+                    Call call = getRemoveMembersFromGroupCall(context, this.tries() + 1,
+                        nextAllowedTryMs);
+                    rescheduleFindCoordinatorTask(context, () -> call);
```
I think there's a reasonable way to consolidate the bookkeeping in `rescheduleFindCoordinatorTask`. If we add the failed call to the argument list, then we can do something like this:
```java
    private void rescheduleFindCoordinatorTask(ConsumerGroupOperationContext<?, ?> context,
                                               Call failedCall,
                                               Supplier<Call> nextCall) {
        log.info("Node {} is no longer the Coordinator. Retrying with new coordinator.",
                context.node().orElse(null));
        // Requeue the task so that we can try with new coordinator
        context.setNode(null);

        Call call = nextCall.get();
        call.tries = failedCall.tries + 1;
        call.nextAllowedTryMs = calculateNextAllowedRetryMs();

        if (call.tries > maxRetries) {
            log.debug("Max retries {} for {} reached", maxRetries, call);
            call.fail(time.milliseconds(), new TimeoutException());
            return;
        }
        Call findCoordinatorCall = getFindCoordinatorCall(context, nextCall);
        runnable.call(findCoordinatorCall, time.milliseconds());
    }
```

Then the caller changes to this (for example):

```java
                    Call call = getDescribeConsumerGroupsCall(context);
                    rescheduleFindCoordinatorTask(context, this, () -> call);
```

What do you think?

skaundinya15 replied:
Yeah I think this does make it a lot cleaner and easier to read. I will refactor to change it to this, and experiment with checking the max tries in `enqueue` as well.

hachikuji reviewed (commented):
Thanks for the updates. Just a couple more comments.

skaundinya15 committed:
addressing PR comments

skaundinya15 committed:
removing unnecessary code

skaundinya15 commented:
Thanks for the reviews @hachikuji! I updated the code to address your comments. I had to change the tests a bit to make to accommodate for moving the tries logic into `enqueue` but other than that everything else seems to work okay. Ready for another round of reviews whenever you are.

hachikuji reviewed (approved):
LGTM. Thanks for the patch!

hachikuji commented:
retest this please

hachikuji commented:
retest this please
